### PR TITLE
Flow websocket

### PR DESCRIPTION
**Pull request by lebauce**

### CODE REVIEW

**Comment by nplanel** [comment]
run tests

**Comment by lebauce** [comment]
run scale-tests

File: analyzer/flow_client.go [hunk 1]
```diff
@@ -102,34 +158,58 @@ func (c *FlowClient) SendFlows(flows []*flow.Flow) {
...
 		}
 	}
 
-	p.flowClients = append(p.flowClients, NewFlowClient(c.Addr, c.Port))
+	flowClient, err := NewFlowClient(c.Addr, c.Port)
+	if err != nil {
+		logging.GetLogger().Error(err)
+	}
+
+	p.flowClients = append(p.flowClients, flowClient)
```
Comment (nplanel) [inline analyzer/flow_client.go]: may not append it in case of error

File: analyzer/flow_server.go [hunk 2]
```diff
@@ -23,275 +23,171 @@
...
 	"github.com/skydive-project/skydive/flow/enhancers"
 	"github.com/skydive-project/skydive/flow/storage"
+	shttp "github.com/skydive-project/skydive/http"
 	"github.com/skydive-project/skydive/logging"
 	"github.com/skydive-project/skydive/probe"
 	"github.com/skydive-project/skydive/topology/graph"
 )
 
-var ErrFlowUDPAcceptNotSupported = errors.New("UDP connection is datagram based (not connected), accept() not supported")
+var writeWait = 10 * time.Second
```
Comment (nplanel) [inline analyzer/flow_server.go]: unused

File: http/wsclient.go [hunk 3]
```diff
@@ -398,24 +471,15 @@ func (a *WSAsyncClientPool) DisconnectAll() {
...
+func (a *WSMessageAsyncClientPool) AddEventHandler(h WSClientEventHandler) {
 	a.eventHandlersLock.Lock()
 	a.eventHandlers = append(a.eventHandlers, h)
-
-	// add this handler per namespace
-	for _, ns := range namespaces {
-		if _, ok := a.nsEventHandlers[ns]; !ok {
-			a.nsEventHandlers[ns] = []WSClientEventHandler{h}
-		} else {
-			a.nsEventHandlers[ns] = append(a.nsEventHandlers[ns], h)
```
Comment (nplanel) [inline http/wsclient.go]: I think you broke the wildcard namespace by removing, basically you forgot to set it afterwards

File: analyzer/flow_server.go (hunk 2 above)
Comment (lebauce) [inline analyzer/flow_server.go]: removed

File: analyzer/flow_client.go (hunk 1 above)
Comment (lebauce) [inline analyzer/flow_client.go]: fixed

**Comment by lebauce** [comment]
run unit-tests

**Comment by lebauce** [comment]
run selenium-tests

**Comment by lebauce** [comment]
run unit-tests

**Comment by lebauce** [comment]
run tests

**Comment by lebauce** [comment]
run scale-tests

**Comment by lebauce** [comment]
run tests

**Comment by lebauce** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by lebauce** [comment]
run tests

**Comment by nplanel** [comment]
LGTM

File: common/types.go [hunk 4]
```diff
@@ -43,6 +43,10 @@ var (
 	ErrFieldNotFound = errors.New("Field not found")
 	// ErrFieldWrongType error field has wrong type
 	ErrFieldWrongType = errors.New("Field has wrong type")
+	// ErrNotFound error not result was found
```
Comment (nplanel) [inline common/types.go]: s/not/no/

File: analyzer/flow_client.go [hunk 5]
```diff
@@ -107,36 +175,63 @@ func (c *FlowClient) SendFlows(flows []*flow.Flow) {
...
-func NewFlowClient(addr string, port int) *FlowClient {
-	FlowClient := &FlowClient{Addr: addr, Port: port}
-	FlowClient.connect()
-	return FlowClient
+func NewFlowClient(addr string, port int) (*FlowClient, error) {
+	var (
+		connection FlowClientConn
+		err        error
+	)
+	protocol := strings.ToLower(config.GetConfig().GetString("flow.protocol"))
```
Comment (safchain) [inline analyzer/flow_client.go]: It could be nice to start scale CI twice one with udp one with websocket

File: http/wsclient.go [hunk 6]
```diff
@@ -200,236 +260,115 @@ func (c *WSAsyncClient) connect() {
...
+	if err := c.wsConn.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
+		logging.GetLogger().Warningf("Error while sending ping to the websocket: %s", err.Error())
+	}
+}
+
+// Connect to the server - and reconnect if necessary
 func (c *WSAsyncClient) Connect() {
 	go func() {
-		for c.running.Load() == true {
+		for c.running.Load() == true && c.reconnect {
```
Comment (lebauce) [inline http/wsclient.go]: done

File: http/wsclient.go [hunk 7]
```diff
@@ -39,90 +37,142 @@ import (
...
+func (c *WSAsyncClient) SendMessage(msg []byte) error {
+	c.wsConn.SetWriteDeadline(time.Now().Add(writeWait))
+
 	w, err := c.wsConn.NextWriter(websocket.TextMessage)
 	if err != nil {
 		return err
 	}
 
-	_, err = io.WriteString(w, msg)
+	_, err = io.WriteString(w, string(msg))
```
Comment (lebauce) [inline http/wsclient.go]: done

File: http/wsclient.go [hunk 8]
```diff
@@ -39,90 +37,142 @@ import (
...
-	OnDisconnected(c *WSAsyncClient)
+const (
+	maxMessages    = 1024
+	maxMessageSize = 0
+	writeWait      = 10 * time.Second
+)
+
+// Interface of a message to send over the wire
+type Message interface {
+	String() string
```
Comment (lebauce) [inline http/wsclient.go]: done

File: http/wsmessage.go [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+		}
+		return
+	}
+
+	a.Dispatch(c, msg)
+}
+
+func (c *WSMessageAsyncClient) OnWSMessage(client WSClient, msg WSMessage) {
+	c.replyChanMutex.RLock()
+	ch, ok := c.replyChan[msg.UUID]
```
Comment (lebauce) [inline http/wsmessage.go]: done

File: http/wsmessage.go [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+
+func NewWSMessageServer(server *WSServer) *WSMessageServer {
+	pool := NewWSMessageClientPool(server.WSClientPool)
+	s := &WSMessageServer{
+		pool:     pool,
+		WSServer: server,
+	}
+
+	s.WSServer.clientHandler = func(conn *websocket.Conn, r *auth.AuthenticatedRequest) WSClient {
+		c := NewWSMessageAsyncClient(DefaultClientHandler(conn, r).(*WSAsyncClient))
```
Comment (lebauce) [inline http/wsmessage.go]: done

File: http/master.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	a.Unlock()
+}
+
+// OnDisconnected event
+func (a *WSMasterElection) OnDisconnected(c WSClient) {
+	a.Lock()
+	if a.master != nil && a.master.GetHost() == c.GetHost() {
+		defer func() {
+			a.selectMaster()
```
Comment (lebauce) [inline http/master.go]: done

File: analyzer/flow_client.go (hunk 5 above)
Comment (lebauce) [inline analyzer/flow_client.go]: done (both)

File: analyzer/flow_client.go [hunk 12]
```diff
@@ -42,53 +44,119 @@ type FlowClientPool struct {
...
+	return nil
+}
+
+// Send data over the wire
+func (c *FlowClientWebSocketConn) Send(data []byte) error {
+	c.wsClient.SendMessage(data)
+	return nil
+}
+
+// NewFlowClientUDPConn returns a new WebScoket flow client
```
Comment (lebauce) [inline analyzer/flow_client.go]: done

File: http/wsserver.go [hunk 13]
```diff
@@ -23,494 +23,78 @@
...
+		s.WSClientPool.RLock()
+		for _, c := range s.clients {
+			if c.GetHost() == host {
 				logging.GetLogger().Errorf("host_id error, connection from %s(%s) conflicts with another one", r.RemoteAddr, host)
 				w.WriteHeader(http.StatusConflict)
-				s.RUnlock()
-
-				s.unregister <- c
-
+				s.WSClientPool.RUnlock()
```
Comment (lebauce) [inline http/wsserver.go]: done

**Comment by lebauce** [comment]
run scale-tests

File: http/wsserver_test.go [hunk 14]
```diff
@@ -24,47 +24,54 @@ package http
...
 
-func (f *fakeServerSubscriptionHandler) OnRegisterClient(c *WSClient) {
-	c.SendWSMessage(NewWSMessage("SrvValidNS", "SrvValidNSUnicast1", "AAA", "001"))
-	c.SendWSMessage(NewWSMessage("SrvNotValidNS", "SrvNotValidNSUnicast2", "AAA", "001"))
-	c.SendWSMessage(NewWSMessage("SrvValidNS", "SrvValidNSUnicast3", "AAA", "001"))
-
-	f.server.BroadcastWSMessage(NewWSMessage("SrvValidNS", "SrvValidNSBroadcast1", "AAA", "001"))
-	f.server.BroadcastWSMessage(NewWSMessage("SrvNotValidNS", "SrvNotValidNSBroacast2", "AAA", "001"))
-	f.server.BroadcastWSMessage(NewWSMessage("SrvValidNS", "SrvValidNSBroadcast3", "AAA", "001"))
+func (f *fakeServerSubscriptionHandler) OnConnected(c WSClient) {
```
Comment (lebauce) [inline http/wsserver_test.go]: They are almost the same, true. But having wsserver_test.go makes it easier to debug simple cases

File: http/wsclient.go [hunk 15]
```diff
@@ -39,90 +36,141 @@ import (
...
 }
 
-func (c *WSAsyncClient) send(msg string) error {
+// Add a message to the send queue
+func (c *WSAsyncClient) Send(m Message) {
+	if c.running.Load() == false {
+		return
+	}
+
+	c.send <- []byte(m.Bytes())
```
Comment (safchain) [inline http/wsclient.go]: m.Bytes() only, already returning []byte

File: http/wsclient.go [hunk 16]
```diff
@@ -39,90 +36,141 @@ import (
...
+func (c *WSAsyncClient) SendMessage(msg []byte) error {
+	c.wsConn.SetWriteDeadline(time.Now().Add(writeWait))
+
 	w, err := c.wsConn.NextWriter(websocket.TextMessage)
 	if err != nil {
 		return err
 	}
 
-	_, err = io.WriteString(w, msg)
+	_, err = w.Write(msg)
```
Comment (safchain) [inline http/wsclient.go]: because I know you like that 

!!! one line :)

File: http/wsclient.go [hunk 17]
```diff
@@ -200,33 +258,40 @@ func (c *WSAsyncClient) connect() {
...
-			return
+			c.RUnlock()
+		case <-c.ticker.C:
+			c.SendPing()
 		}
 	}
 }
 
+func (c *WSAsyncClient) SendPing() {
+	c.Lock()
```
Comment (safchain) [inline http/wsclient.go]: not sure you need to lock

File: http/wsserver_test.go (hunk 14 above)
Comment (safchain) [inline http/wsserver_test.go]: ok to keep it

**Comment by safchain** [comment]
You need to rebase it, BTW

File: http/wsclient.go (hunk 17 above)
Comment (lebauce) [inline http/wsclient.go]: done

File: config/config.go [hunk 18]
```diff
@@ -51,60 +51,79 @@ func init() {
...
+
+	cfg.SetDefault("docker.url", "unix:///var/run/docker.sock")
+
+	cfg.SetDefault("etcd.data_dir", "/var/lib/skydive/etcd")
+	cfg.SetDefault("etcd.embedded", true)
+	cfg.SetDefault("etcd.listen", "localhost:2379")
+
+	cfg.SetDefault("flow.expire", 600)
+	cfg.SetDefault("flow.update", 60)
+	cfg.SetDefault("flow.protocol", "udp")
```
Comment (lebauce) [inline config/config.go]: indeed

File: agent/topology_forwarder.go [hunk 19]
```diff
@@ -47,79 +45,71 @@ func (t *TopologyForwarder) triggerResync() {
...
-// OnDisconnected websocket event handler
-func (t *TopologyForwarder) OnDisconnected(c *shttp.WSAsyncClient) {
-	if c == t.master {
-		t.master = t.WSAsyncClientPool.MasterClient()
-
-		// re-sync as we changed of master and some message could have lost by the previous one
+// OnNewMaster websocket event handler
+func (t *TopologyForwarder) OnNewMaster(c shttp.WSClient) {
+	if c == nil {
+		logging.GetLogger().Infof("Lost connection to master")
```
Comment (lebauce) [inline agent/topology_forwarder.go]: right

File: Makefile [hunk 20]
```diff
@@ -5,6 +5,7 @@ export GO15VENDOREXPERIMENT=1
 
 FLOW_PROTO_FILES=flow/flow.proto flow/set.proto flow/request.proto
 FILTERS_PROTO_FILES=filters/filters.proto
+FLOW_PROTOCOL?=websocket
```
Comment (lebauce) [inline Makefile]: removed

File: http/wsclient.go (hunk 15 above)
Comment (lebauce) [inline http/wsclient.go]: done

**Comment by safchain** [comment]
rerun functional-tests-backend-orientdb

**Comment by safchain** [comment]
rerun functional-tests-backend-orientdb

**Comment by lebauce** [comment]
run unit-tests

**Pull request merged**